src/overlays/actors/ovl_Bg_Po_Event/z_bg_po_event.c

```diff
@@ -21,47 +48,601 @@ const ActorInit Bg_Po_Event_InitVars = {
+s32 BgPoEvent_NextPicture(BgPoEvent* this) {
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Po_Event/func_808A69DC.s")
+    if ((this->dyna.actor.parent != NULL) && (this->dyna.actor.child != NULL)) {
+        if (Math_Rand_ZeroOne() < 0.5f) {
+            puzzleState = ((BgPoEvent*)this->dyna.actor.parent)->actorIndex;
+        } else {
+            puzzleState = ((BgPoEvent*)this->dyna.actor.child)->actorIndex;
+        }
+    } else if (this->dyna.actor.parent != NULL) {
+        puzzleState = ((BgPoEvent*)this->dyna.actor.parent)->actorIndex;
+    } else if (this->dyna.actor.child != NULL) {
+        puzzleState = ((BgPoEvent*)this->dyna.actor.child)->actorIndex;
+    } else {
+        return 0;
+    }
+    return 1;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Po_Event/func_808A6A94.s")
+void BgPoEvent_PaintingEmpty(BgPoEvent* this, GlobalContext* globalCtx) {
+    if (puzzleState == this->actorIndex) {
+        this->timer = 255;
+        this->actionFunc = BgPoEvent_PaintingAppear;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Po_Event/func_808A6BC8.s")
+void BgPoEvent_PaintingAppear(BgPoEvent* this, GlobalContext* globalCtx) {
+    this->timer -= 20;
+    if (this->timer <= 0) {
+        this->timer = 1000;
+        this->actionFunc = BgPoEvent_PaintingPresent;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Po_Event/func_808A6CCC.s")
+void BgPoEvent_PaintingVanish(BgPoEvent* this, GlobalContext* globalCtx) {
+    this->timer += 20;
+    if (this->timer >= 255) {
+        BgPoEvent_NextPicture(this);
+        this->actionFunc = BgPoEvent_PaintingEmpty;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Po_Event/func_808A6F7C.s")
+void BgPoEvent_PaintingPresent(BgPoEvent* this, GlobalContext* globalCtx) {
+    Actor* thisx = &this->dyna.actor;
+    Player* player = PLAYER;
+    s32 phi_v0;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Po_Event/func_808A7138.s")
+    DECR(this->timer);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Po_Event/func_808A7238.s")
+    if ((this->timer == 0) || ((thisx->xzDistFromLink < 150.0f) && (thisx->yDistFromLink < 50.0f)) ||
+        (((func_8002DD78(player) != 0) && (thisx->xzDistFromLink < 320.0f)) &&
+         ((this->actorIndex != 2) ? (thisx->yDistFromLink < 100.0f) : (thisx->yDistFromLink < 0.0f)) &&
+         (func_8002DFC8(thisx, 0x2000, globalCtx) != 0))) {
+        /*This last conditional is checking if
+            1) Link is holding a ranged weapon
+            2) Link is too close in the xz plane
+            3) Link is too close in the y direction. The painting
+               under the balcony allows him to be closer.
+            4) Link is within 45 degrees of facing the painting. */
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Po_Event/func_808A72AC.s")
+        if ((thisx->parent != NULL) || (thisx->child != NULL)) {
```

Comment:
can you merge this if into its parent? making an even bigger condition 😄 
It would allow to remove the return and wrap the rest of the function in an `else` clause though

Reply:
Yes, and that's probably how it was written. super ugly.